### DISCUSSION TITLE
Memory/Handlers Leak Issues: Rendering Tree of CustomElement using only lit-html

### DISCUSSION THREAD

**Discussion by fopsdev** [lit/lit#2977] [General]
Hi there

I'm using lit-html successfully for quite a few years now in my own framework.
I'm currently on lit-html 2.0.2 buit i've also tried with most recent version.
I'm using it inside my own Custom Elements (Directly inherited from HtmlElement)

Updated:
It happens when just render a literal that has a undefined e.g. ${myComp} in it
See the repro further down.

My question is: Does lit-html keeps some references in that case? Because i can see the used handlers going up in devtools.
(And yes, i'm forcing the Garbage Collector from inside devtools)

I would be happy for some pointers from top of your head.

(And i'll use Lit-Element maybe one day but currently i'm maintaining quite a few projects using this approach.)

**Comment by justinfagnani** [comment]
Any references lit-html maintains are kept on the elements themselves (like [here](https://github.com/lit/lit/blob/main/packages/lit-html/src/lit-html.ts#L632)), which will be cleaned up when the elements are collected.

I could looks at a simple reproduction to see if there's something going on though.

**Reply by fopsdev** [reply]
thanks for chiming in here.
yes i saw this __$litPart$_ on the element itself.
i was even (just for testing) setting this.__$litPart$_ = null in my disconnectedCallback of the custom element but it didn't help
let me come up with a repro then (can take some time)

**Comment by fopsdev** [comment]
@justinfagnani 
i couldn't come up with a simple reproduction because it's quite a powerful system (basically using js proxies to detect spots on the page which needs to be rerendered)

never the less i've removed the fluff from my components.

This hopefully helps you to spot if i'm doing something fundamentally wrong

https://gist.github.com/fopsdev/64fc98350c50c75b5a7de394b0a4453c

**Reply by fopsdev** [reply]
after opening and closing the login dialog i have the following detached elements:

![image](https://user-images.githubusercontent.com/5448300/171673571-301a0da0-de05-4431-a4c5-e68a997852e7.png)

maybe it helps

**Comment by kevinpschaaf** [comment]
It is not safe to manually remove nodes rendered by lit-html _inside_ the container you passed to `render`. From your code, if `//remove screen from dom` is removing a node that was rendered in the container via a lit-html template, it's very possible that that node (or a child of it) is still retained by lit in the `$litPart$` tracking object stored on the container node.

As a rule of thumb, if Lit rendered the node, you should also use Lit to remove it. Typically this is done via a conditional in the template based on state, aka:

```
html`<div>${shouldRenderPopup ? html`<div class=".popup"></div>` : nothing}`
```

So rather than directly calling `popup.parentNode.removeChild(popup)`, you should re-render the template, passing `shouldRenderPopup` as false.

**Reply by fopsdev** [reply]
ok, thanks i'll check if that could be the issue.
btw. i have already tried to set ```$litPart$``` to null in the disconnected callback but at no avail
and as i mentioned in my gist:
```
    // i have some further logic which adds/removes the screens to the dom (check the divs in index.html)
    // i'm doing this because i can setup my screens in a config part of my application and there could be dozens of screens
    // and i don't want to do smth like:
    // let screen
    // if (currentScreen === "Dashboard") {
    //   screen = html`<my-dashboard></my-dashboard>`
    // }
    // else if (currentScreen === "Feature1") {
    //     screen =  html`<my-feature1></my-feature1>`
    // }
    // ...
    // ...
    // return screen
    //
    // other than that the rendering is state based as it should be imo
```
i don't like the conditional approach for this usecase.

**Reply by fopsdev** [reply]
ok i switched from my dom-removing stuff for that particular case to a conditional:
```
let loginDialog
    if (this.state.ovl.dialogs.Login.visible) {
      loginDialog = html`<ovl-login></ovl-login>`
    }
    app = html`
      <div
        id="OvlScreens"
        class="${scrollable}"
        @click=${handleTopLevelClick}
      ></div>
      <div id="OvlDialogs">${loginDialog}</div>
      <ovl-snack> </ovl-snack>
    `
```
but it didn't help

**Reply by justinfagnani** [reply]
> btw. i have already tried to set $litPart$ to null in the disconnected callback but at no avail

What are you setting `$litPart$` to null on? First, you shouldn't ever have to do that. If you called `render(myTemplate)` to render something, just call `render(undefined)` to render nothing. Second, for a LitElement the render container is the ShadowRoot. Were you modifying that? Third, it seems like you're dynamically adding elements to the DOM yourself? In that case Lit didn't render them and there's no node that it matters to mess with `$litPart$` on.

**Reply by fopsdev** [reply]
I've set ```$litPart$``` to null just as a test. Basically just to check if it can help to get rid of the retainers. I know its bad :)
I'm not using lit-element or shadow dom at all in my components. 
Yes i'm adding DOM Elements myself. But they contain e.g. a ```<my-login></my-login>``` component whicht itself then starts the element lifecycle and therefore uses lit-html to render itself

**Reply by kevinpschaaf** [reply]
I think we'll need you to make a working reproduction to help you dig further into what's going on.

**Reply by fopsdev** [reply]
Ok, thanks for your help guys its highly appreciated. Lemme try to get a small repro project out.

**Comment by fopsdev** [comment]
Ok, made a small repro.
https://github.com/fopsdev/OvlReproMemoryLeak

As you can see there are retainers:
![image](https://user-images.githubusercontent.com/5448300/172358082-73e864f3-d07e-4d5b-8c37-da38fde5c4e9.png)

Thanks for any feedback

**Reply by fopsdev** [reply]
Btw. and just in case: I'm hunting after Retainers using Edge Devtools. I'm not aware of a better Tool yet.
A little Screencast:


https://user-images.githubusercontent.com/5448300/172675522-540d9b9a-f518-4c2f-8f61-70bf60e2cab5.mp4

**Reply by fopsdev** [reply]
I have changed the title because as we can see in the repro it has nothing todo with removing dom-nodes

**Reply by fopsdev** [reply]
An important fact i've just discovered now:

It works with lit-html 1.1.2 -> there are no retainers

**Reply by fopsdev** [reply]
@justinfagnani  lemme know if i should create an issue out of this.

**Comment by fopsdev** [comment]
ping @justinfagnani

**Comment by fopsdev** [comment]
ok, no response here. will create an issue and close this one
